src/formatter/formatter.vala

```diff
@@ -0,0 +1,278 @@
+        var new_file = File.new_build_filename (Environment.get_tmp_dir (), "vls-format%lld.vala".printf (time));
+        try {
+            original.copy (new_file, FileCopyFlags.ALL_METADATA | FileCopyFlags.OVERWRITE, null, null);
+        } catch (Error e) {
+            return "Error copying file: %s".printf (e.message);
+        }
+        string sout;
+        string serr;
+        int exit;
+        try {
+            var env_vars = Environ.get ();
+            for (var i = 0; i < env_vars.length; i++) {
+                var env = env_vars[i];
+                // Patch the PATH variable to include some other "standard paths"
+                if (env.has_prefix ("PATH=") && env.contains ("/usr/bin")) {
+                    env_vars[i] = env + ":/usr/local/bin";
+                }
+            }
+            var spawn_args = new string[] { "uncrustify", "-c", cfg.get_path (), "--replace", "--no-backup", new_file.get_path () };
+            Process.spawn_sync (Environment.get_current_dir (),
+                                spawn_args,
+                                env_vars,
+                                SpawnFlags.SEARCH_PATH_FROM_ENVP,
+                                null,
+                                out sout,
+                                out serr,
+                                out exit);
+        } catch (SpawnError e) {
+            return "Error spawning uncrustify: %s (PATH=%s)".printf (e.message, Environment.get_variable ("PATH"));
+        }
+        if (exit != 0) {
+            return "Uncrustify failed: %s".printf (serr);
+        }
+        var str = "";
+        var n = 0;
+        try {
+            var reopened = File.new_for_path (new_file.get_path ());
+            var iostream = reopened.open_readwrite ();
+            var distream = new DataInputStream (iostream.input_stream);
+            var sb = new StringBuilder ();
+            string tmp = null;
+            size_t len;
+            while ((tmp = distream.read_line (out len)) != null) {
+                sb.append (tmp).append_c ('\n');
+                n++;
+            }
+            str = sb.str;
+        } catch (Error e) {
+            return "Error reading uncrustified file: %s".printf (e.message);
+        }
+
+        edit = new Lsp.TextEdit () {
+            range = new Lsp.Range () {
+                start = new Lsp.Position () {
+                    line = 0,
+                    character = 0
+                },
+                end = new Lsp.Position () {
+                    line = n + 1,
+                    // Just for the trailing newline
+                    character = 1
+                }
+            },
+            newText = str
+        };
+        try {
+            new_file.@delete ();
+            cfg.@delete ();
+        } catch (Error e) {
+            warning ("Error cleaning up: %s", e.message);
+        }
+        return null;
+    }
+
+    File generate_uncrustify_config () throws Error {
```

Comment:
I think we should build an argument list from these options (see uncrustify's `--set <option>=<value>` flag) rather than writing to a config file. Again, this avoids unnecessary I/O.

Reply:
I did it a bit different, that would only require 1 IO-Operation.

Before:
- Write .cfg file (vala-language-server)
- Read .cfg file (uncrustify)

Now:
- Write .cfg file, only if it does not exist (We can derive a hash value from the formatting options)
- Read .cfg file (uncrustify)

I had difficulties with the line `conf["set func_call_user"] = "C_ NC_ N_ Q_ _";`, as I wasn't able to get this to work, everything else would work fine with `--set <key>=<value>`

Reply:
Are you building the argument list as an array of strings and passing it into `SubprocessLauncher.spawnv()`?

Reply:
Currently I use `spawn`, but when I've attempted without any config on the disk I used `spawnv`.